stack/maxima/contrib/linearalgebra.mac

```diff
@@ -0,0 +1,885 @@
+
+/** 
+ * Predicate to determine whether a matrix is in row echelon form
+ * i.e. Is every zero-row below all non-zero rows and does the pivot in each row appear to the right of the previous pivot?
+ * Optionally checks whether each pivot is equal to 1
+ * No RREFp function is given, as rref(M) is unique
+ *
+ * @param[matrix] M An mxn matrix
+ * @param[boolean] normalise_pivots Optional: If true is given here, then pivots are required to be equal to 1
+ * @return[boolean] Is this matrix in row echelon form?
+ */
+REFp(M,[normalise_pivots]):= block([isREF,pivot_row,m,n,jj,ii],
+  if emptyp(normalise_pivots) then normalise_pivots: false else normalise_pivots: first(normalise_pivots),
+  isREF: matrixp(M),
+  if isREF then block(
+    pivot_row: 0,
+    [m, n]: matrix_size(M),
+    for jj: 1 thru n do block(
+      jj: ev(jj,simp),
+      if is(pivot_row < m) then block(
+        if is(M[ev(pivot_row+1,simp),jj] # 0) then block(
+          pivot_row: ev(pivot_row + 1,simp),
+          if normalise_pivots and is(M[ev(pivot_row,simp),jj] # 1) then isREF: false
+        ),
+        for ii: ev(pivot_row+1,simp) thru m do block(
+          ii: ev(ii,simp),
+          if is(M[ii,jj] # 0) then isREF: false
+        )
+      )
+    )
+  ),
+  return(isREF)
+);
+
+/** 
+ * Is a given object a square matrix?
+ * i.e. Does the matrix have the same number of rows as columns?
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Is M a square matrix?
+ */
+squarep(M):= block([isSquare],
+  isSquare: false,
+  if matrixp(M) then block(
+    if is(apply("=",matrix_size(M))) then isSquare: true
+  ),
+  return(isSquare)
+);
+
+/** 
+ * Is a given object a diagonalisable matrix?
+ * i.e. Is there an invertible matrix P and diagonal matrix D such that P^^-1 . M . P = D?
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Is M a diagonalisable matrix?
+ */
+diagonalisablep(M):= if squarep(M) then ev(diagp(dispJordan(jordan(M))),simp) else false;
+
+/** 
+ * Is a given object a symmetric matrix?
+ * i.e. Is M = M^T? 
+ * NOTE: The native function symmetricp() does the same thing and more, but is currently banned.
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Is M a symmetric matrix?
+ */
+sym_p(M):= if squarep(M) then is(M = ev(transpose(M),simp)) else false; 
+
+/** 
+ * Is a given object an invertible matrix?
+ * i.e. Is there a matrix M^^-1 such that M^^-1 . M = M . M^^-1 = I?
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Is M an invertible matrix?
+ */
+invertiblep(M):= block([isInvertible],
+  isInvertible: false,
+  if squarep(M) then block(
+    if ev(is(determinant(M)#0),simp) then isInvertible: true
+  ),
+  return(isInvertible)
+);
+
+/** 
+ * Is a given object a matrix with orthogonal columns?
+ * i.e. for any two distinct columns v1 and v2, is v1 . v2 = 0?
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Does M have orthogonal columns?
+ */
+orthogonal_columnsp(M):= ev(diagp(transpose(M).M),simp);
```

Comment:
Shouldn't the `diagp` be outside the `ev`?

I also noticed that sometimes, even with simp, complicated expressions (with square roots as they may appear in the context of orthonormal matrices) sometimes don't get fully simplified. For example, 
![image](https://github.com/maths/moodle-qtype_stack/assets/678199/a0ddab4c-85b4-450c-90c7-6721956947c5) only collapses into the identity matrix when applying `expand`. I wonder if it's worthwhile adding this?

Reply:
I _think_ that the `ev` is fine in either position. I got a bit paranoid about simplification when writing, as I kept finding that certain functions or expressions were returning unevaluated unexpectedly. My interpretation of this particular expression is "take the expression `diagp(transpose(M).M)` and evaluate it with simplification turned on", so the `transpose(M).M` should still be evaluated as expected before the `diagp` predicate runs. My very quick check now suggests that both are fine, but if I see a counter-example I'm very happy to swap them. 

Yes, that's a good point on `expand`. Could you please send me the problem case you've screenshotted above? I'll need to figure out which functions will need that and that looks like a good test case to use.  

Thanks for your help with little project Georg! It's great to have a second pair of eyes on this :)

Reply:
Fair point about the `ev`, my intuition was that `diagp` returns a boolean anyway so I thought you'd want to simplify the expression within `diagp`, but I guess technically it might as well not evaluate the if-clauses and return them as is, for some reason, and if simplification is on, it'll apply recursively everywhere so no harm putting it on the outside? You probably have a better idea of how all of this works than I do...

The use case where I encountered this simplification issue is actually slightly more complex as I'm using a custom scalar product. I've been trying to reproduce the issue for the standard scalar product, but so far without success.
Anyway, below is some maxima code that you can run to see that expand sometimes does make a difference.

```
/* Compute orthogonal projection with respect to a custom scalar product */
proj(u, v, sp) := sp(v,u)/sp(u,u)*u;
/* Gram-Schmidt orthogonalization with respect to a custom scalar product */
gram_schmidt(M, sp) := block([u_1, u_2,u_3,v_2,v_3,AA],
    /*Prepare Gram-Schmidt process*/
    u_1: col(M,1),
    v_2: col(M,2),
    v_3: col(M,3),
    /*orthogonalize*/
    u_2: v_2 - proj(u_1, v_2, sp),
    u_3: v_3 - proj(u_1, v_3, sp) - proj(u_2, v_3, sp),
    /*normalize*/
    u_1: u_1/sqrt(sp(u_1,u_1)),
    u_2: u_2/sqrt(sp(u_2,u_2)),
    u_3: u_3/sqrt(sp(u_3,u_3)),
    AA: addcol(u_1, u_2, u_3),
    return(AA)
);
/* A symmetric positive definite matrix and corresponding bilinear form (scalar product) */
A: matrix([3,-1,0],[-1,3,0],[0,0,2]);
bform(w,v) := transpose(w).A.v;
/* Orthogonalize the standard basis with respect to this scalar product */
baseM: ident(3);
OB: expand(gram_schmidt(baseM, bform));
/* Customized predicate functions */
orthogonal_columnsp(M, spMat):= diagp(ev(transpose(M).spMat.M,simp));
orthonormal_columnsp(M, spMat):= if matrixp(M) then is(ev(transpose(M).spMat.M,simp) = ident(second(matrix_size(M)))) else false;
/* Check outputs */
orthogonal_columnsp(OB, A);
orthonormal_columnsp(OB, A);
ev(transpose(OB).A.OB, simp);
ev(expand(transpose(OB).ev(expand(A.OB), simp)), simp);
```